### PR TITLE
Editor: Serve as a proxy for the interface package

### PR DESCRIPTION
**Pull request by youknowriad**
Related #52632 

## What?

As you align editor APIs between post and site editors, we'll have to align extensibility APIs that rely on the interface package.

This package is a bundled package (bundled within the package that import it) and it also defines a store, which means it can't be imported both from "editor" and "edit-post" at the same time. (Triggers an error about double store).
I think having a "store" in this package was an error that we made a long time ago, so it's hard to change now in one go, so in order to avoid this issue, we can make the editor package work as a proxy for the "interface" package APIs for both edit-post and edit-site.

So the current PR does exactly that to allow us to make further unification as follow-ups.

## Testing Instructions

There should be no functional change, it's mostly about moving things around.

### CODE REVIEW

File: packages/docgen/lib/get-intermediate-representation.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ const hasVariableWithName = ( node, name ) =>
 	node.declarations.some( ( declaration ) => {
 		if ( declaration.id.type === 'ObjectPattern' ) {
 			return declaration.id.properties.some(
-				( property ) => property.key.name === name
+				( property ) => property.key && property.key.name === name
```
Comment (jsnajdr) [inline packages/docgen/lib/get-intermediate-representation.js]: If you want to catch both `a` and `b` in a `const { a, ...b } = ...` declaration, you need to check:
```js
property.key?.name === name || property.argument?.name === name
```
The first one is for normal keys like `a`, the second one for rest keys like `b`. Here is what ASTExplorer shows for `const { a, ...b }` syntax:

Screenshot 2024-04-16 at 10 59 35

Comment (youknowriad) [reply packages/docgen/lib/get-intermediate-representation.js]: I'm not entirely sure what this function does to be honest, but I can make that change :P

Comment (ntsekouras) [reply packages/docgen/lib/get-intermediate-representation.js]: That was interesting!

Comment (jsnajdr) [reply packages/docgen/lib/get-intermediate-representation.js]: I think it's used to find a connection between an exported name and a JSDoc comment. For example:
```js
/**
 * Some comment with documentation about `b`.
 */
const { a, ...b } = foo();

export { b };
```
Here the comment will be used as a documentation for the exported name `b`.

**Review by ntsekouras — APPROVED** [review]
I smoke tested the editors and didn't notice anything different.

File: packages/docgen/lib/get-intermediate-representation.js (hunk 1 above)
Comment (oandregal) [reply packages/docgen/lib/get-intermediate-representation.js]: Pushed a test for this https://github.com/WordPress/gutenberg/pull/60748/commits/2c7526dbe494bedbc52ac023b510bf02138f6ee4 but Jarda beat me to the suggestion :)

Comment (oandregal) [reply packages/docgen/lib/get-intermediate-representation.js]: I've got another test to add, I've disabled auto-merge. I'll merge after.

**Pull request merged**